content/docs/04-clusters/03-edge/30-cloud-init.md

```diff
@@ -0,0 +1,287 @@
+| `before-upgrade`    | This stage is executed before upgrade                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |      |
+| `before-reset`      | This stage is executed before reset                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |      |
+
+
+<InfoBox>
+
+ Each stage has a before and after hook you can use to achieve more granular customization. For example, you can use `network.after` to verify network connectivity.
+
+ </InfoBox>
+
+
+# Where to apply Cloud-Init Stages?
```

Comment:
I would complete this section by adding that "all user-data is preserved to final stage." as Rishi mentioned

Reply:
@cloudmaniac I do not disagree with your point, but help me understand here. How does that help the user, or in what context would the user need that information? I'm mainly trying to identify if the information would be noise or if there is an actual need.

Reply:
@cloudmaniac How would they compete? If you use, for example, the stage `boot` the installer and the OS pack, each stage will run when applicable. Whatever you do in the installer stage will happen in the installer stage, and what you do at the OS stage happens when the OS is booting up.

Reply:
For example if an admin create 2 incompatible/competing configuration in the same stage (e.g., `fs` or `network`), what happens?

Reply:
it is worth noting (and please check with Rishi), that any configuration you execute in a later stage could overrwrite something done in an earlier stage. But of course if there are different boots (installer, OS) both running cloud init, then they would be like two separate namespaces and could only overwrite what had happened in their own namespaces.

Reply:
@rishi-anand any chance we could get your input on the discussion above?